pkg/agent/route/route_linux.go

```diff
@@ -1344,51 +1341,57 @@ func (c *Client) AddClusterIPRoute(svcIP net.IP) error {
-		newClusterIPCIDR = &net.IPNet{IP: svcIP, Mask: net.CIDRMask(mask, mask)}
+	// Enlarge the ClusterIP CIDR with the ClusterIP. Note that, if the ClusterIP CIDR doesn't exist, just generate a CIDR
+	// with the ClusterIP.
+	newClusterIPCIDR, err := util.ExtendCIDRWithIP(curClusterIPCIDR, svcIP)
+	if err != nil {
+		return fmt.Errorf("enlarge the destination CIDR with an error: %w", err)
 	}
 
-	// Generate a route with the new destination CIDR and install it.
+	// Generate a route with the new ClusterIP CIDR and install it.
 	newClusterIPCIDRMask, _ := newClusterIPCIDR.Mask.Size()
 	route := generateRoute(newClusterIPCIDR.IP, newClusterIPCIDRMask, gw, linkIndex, scope)
 	if err = c.netlink.RouteReplace(route); err != nil {
 		return fmt.Errorf("failed to install new ClusterIP route: %w", err)
 	}
-	// Store the new destination CIDR.
+
+	// Store the new ClusterIP CIDR and the new generated route to serviceRoutes. Then the route can be restored by
+	// periodical sync if it is deleted by accident.
 	if isIPv6 {
 		c.clusterIPv6CIDR = route.Dst
+		c.serviceRoutes.Store(clusterIPv6CIDRKey, route)
 	} else {
 		c.clusterIPv4CIDR = route.Dst
+		c.serviceRoutes.Store(clusterIPv4CIDRKey, route)
 	}
 	klog.V(4).InfoS("Created a route to route the ClusterIP to Antrea gateway", "route", route, "ClusterIP", svcIP)
 
 	// Collect stale routes.
 	var staleRoutes []*netlink.Route
 	if curClusterIPCIDR != nil {
-		// If current destination CIDR is not nil, the route with current destination CIDR should be uninstalled.
+		// If the current ClusterIP CIDR exists, the route with the current ClusterIP CIDR should be uninstalled since
+		// a new route with a newly calculated ClusterIP CIDR has been installed.
 		route.Dst = curClusterIPCIDR
 		staleRoutes = []*netlink.Route{route}
 	} else {
-		// If current destination CIDR is nil, which means that Antrea Agent has just started, then all existing routes
-		// whose destination CIDR contains the first ClusterIP should be uninstalled, except the newly installed route.
-		// Note that, there may be multiple stale routes prior to this commit. When upgrading, all stale routes will be
-		// collected. After this commit, there will be only one stale route after Antrea Agent started.
 		routes, err := c.listIPRoutesOnGW()
 		if err != nil {
 			return fmt.Errorf("error listing ip routes: %w", err)
 		}
+		// If current ClusterIP CIDR doesn't exist, which means that Antrea Agent has just started, and the routes for
+		// ClusterIP generated in the last round should be deleted.
 		for i := 0; i < len(routes); i++ {
-			if routes[i].Gw.Equal(gw) && !routes[i].Dst.IP.Equal(svcIP) && routes[i].Dst.Contains(svcIP) {
+			// Skip the route whose destination CIDR has only one address since it could be the route for the first
+			// ClusterIP newly installed. Note that, this could also match a route for a LoadBalancer ingress IP, but
+			// it doesn't affect skipping the route for the first ClusterIP.
```

Comment:
Should it just check the route doesn't match the new route?

Reply:
This if branch only matches the installation of the first ClusterIP, and the netmask of the route for the first ClusterIP is 32 or 128, we don't need to check that if the route matches the new added route.

Reply:
I meant it could handle all cases instead of just one specific case. The code would also match more the purpose and don't need to explain ClusterIP case.
```
for i := 0; i < len(routes); i++ {
    // Ignore if it's not for cluster IPs.
    if !routes[i].Dst.Contains(svcIP) || !routes[i].Gw.Equal(gw) {
        continue
    }
    // Ignore if it's the expected route.
    if routes[i].Equal(route) {
        continue
    }
    if err = c.netlink.RouteDel(rt); err != nil {
    ...
}
```
Isn't above more clear?

Reply:
Much better.